// test/hornet-js-test/test/hornet-test-assert-karma.js
"use strict";
/**
 * Copyright ou © ou Copr. Ministère de l'Europe et des Affaires étrangères (2017)
 * <p/>
 * pole-architecture.dga-dsi-psi@diplomatie.gouv.fr
 * <p/>
 * Ce logiciel est un programme informatique servant à faciliter la création
 * d'applications Web conformément aux référentiels généraux français : RGI, RGS et RGAA
 * <p/>
 * Ce logiciel est régi par la licence CeCILL soumise au droit français et
 * respectant les principes de diffusion des logiciels libres. Vous pouvez
 * utiliser, modifier et/ou redistribuer ce programme sous les conditions
 * de la licence CeCILL telle que diffusée par le CEA, le CNRS et l'INRIA
 * sur le site "http://www.cecill.info".
 * <p/>
 * En contrepartie de l'accessibilité au code source et des droits de copie,
 * de modification et de redistribution accordés par cette licence, il n'est
 * offert aux utilisateurs qu'une garantie limitée.  Pour les mêmes raisons,
 * seule une responsabilité restreinte pèse sur l'auteur du programme,  le
 * titulaire des droits patrimoniaux et les concédants successifs.
 * <p/>
 * A cet égard  l'attention de l'utilisateur est attirée sur les risques
 * associés au chargement,  à l'utilisation,  à la modification et/ou au
 * développement et à la reproduction du logiciel par l'utilisateur étant
 * donné sa spécificité de logiciel libre, qui peut le rendre complexe à
 * manipuler et qui le réserve donc à des développeurs et des professionnels
 * avertis possédant  des  connaissances  informatiques approfondies.  Les
 * utilisateurs sont donc invités à charger  et  tester  l'adéquation  du
 * logiciel à leurs besoins dans des conditions permettant d'assurer la
 * sécurité de leurs systèmes et ou de leurs données et, plus généralement,
 * à l'utiliser et l'exploiter dans les mêmes conditions de sécurité.
 * <p/>
 * Le fait que vous puissiez accéder à cet en-tête signifie que vous avez
 * pris connaissance de la licence CeCILL, et que vous en avez accepté les
 * termes.
 * <p/>
 * <p/>
 * Copyright or © or Copr. Ministry for Europe and Foreign Affairs (2017)
 * <p/>
 * pole-architecture.dga-dsi-psi@diplomatie.gouv.fr
 * <p/>
 * This software is a computer program whose purpose is to facilitate creation of
 * web application in accordance with french general repositories : RGI, RGS and RGAA.
 * <p/>
 * This software is governed by the CeCILL license under French law and
 * abiding by the rules of distribution of free software.  You can  use,
 * modify and/ or redistribute the software under the terms of the CeCILL
 * license as circulated by CEA, CNRS and INRIA at the following URL
 * "http://www.cecill.info".
 * <p/>
 * As a counterpart to the access to the source code and  rights to copy,
 * modify and redistribute granted by the license, users are provided only
 * with a limited warranty  and the software's author,  the holder of the
 * economic rights,  and the successive licensors  have only  limited
 * liability.
 * <p/>
 * In this respect, the user's attention is drawn to the risks associated
 * with loading,  using,  modifying and/or developing or reproducing the
 * software by the user in light of its specific status of free software,
 * that may mean  that it is complicated to manipulate,  and  that  also
 * therefore means  that it is reserved for developers  and  experienced
 * professionals having in-depth computer knowledge. Users are therefore
 * encouraged to load and test the software's suitability as regards their
 * requirements in conditions enabling the security of their systems and/or
 * data to be ensured and,  more generally, to use and operate it in the
 * same conditions as regards security.
 * <p/>
 * The fact that you are presently reading this means that you have had
 * knowledge of the CeCILL license and that you accept its terms.
 *
 */
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
/**
 * hornet-js-react-components - Ensemble des composants web React de base de hornet-js
 *
 * @author MEAE - Ministère de l'Europe et des Affaires étrangères
 * @version v5.1.1
 * @link git+https://github.com/diplomatiegouvfr/hornet-js.git
 * @license CECILL-2.1
 */
var test_run_1 = require("hornet-js-test/src/test-run");
var decorators_1 = require("hornet-js-test/src/decorators");
var hornet_test_assert_1 = require("hornet-js-test/src/hornet-test-assert");
var base_test_1 = require("hornet-js-test/src/base-test");
/**
 * Classe de test pour HornetTestAssert
 */
var HornetTestAssertKarma = /** @class */ (function (_super) {
    tslib_1.__extends(HornetTestAssertKarma, _super);
    function HornetTestAssertKarma() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    HornetTestAssertKarma.prototype.assertEqualsTest = function () {
        var value1 = "titi";
        var expected1 = "titi";
        hornet_test_assert_1.HornetTestAssert.assertEquals(expected1, value1, "Les chaines de caractères identiques ne sont pas considérées comme tel");
        var value2 = 2;
        var expected2 = 2;
        hornet_test_assert_1.HornetTestAssert.assertEquals(expected2, value2, "Les numériques identiques ne sont pas considérés comme tel");
        var value3 = { label: "titi", value: 2 };
        var expected3 = { label: "titi", value: 2 };
        hornet_test_assert_1.HornetTestAssert.assertEquals(expected3, value3, "Les objets identiques ne sont pas considérés comme tel");
        this.end();
    };
    ;
    HornetTestAssertKarma.prototype.assertNotEqualsTest = function () {
        var value1 = "titi";
        var expected1 = "toto";
        var value2 = 2;
        var expected2 = 22;
        var value3 = { label: "titi", value: 2 };
        var expected3 = { label: "titi", value: 44 };
        hornet_test_assert_1.HornetTestAssert.assertNotEquals(expected1, value1, "Les chaines de caractères différentes sont considérées comme identiques");
        hornet_test_assert_1.HornetTestAssert.assertNotEquals(expected2, value2, "Les numériques différents sont considérés comme identiques");
        hornet_test_assert_1.HornetTestAssert.assertNotEquals(expected3, value3, "Les objets différents sont considérés comme identique");
        this.end();
    };
    ;
    HornetTestAssertKarma.prototype.assertNullTest = function () {
        hornet_test_assert_1.HornetTestAssert.assertNull(null, "null n'est pas considéré comme null");
        this.end();
    };
    ;
    HornetTestAssertKarma.prototype.assertNotNullTest = function () {
        hornet_test_assert_1.HornetTestAssert.assertNotNull(1, "1 est comme considéré comme null");
        hornet_test_assert_1.HornetTestAssert.assertNotNull("titi", "la chaîne de caractère est comme considérée comme null");
        hornet_test_assert_1.HornetTestAssert.assertNotNull({ label: "titi", value: 2 }, "l'objet est comme considérée comme null");
        this.end();
    };
    ;
    HornetTestAssertKarma.prototype.assertTrueTest = function () {
        var obj = { label: "titi", value: 2 };
        hornet_test_assert_1.HornetTestAssert.assertTrue(true, "true n'est pas considéré comme true");
        hornet_test_assert_1.HornetTestAssert.assertTrue(1 == 1, "1 == 1 n'est pas considéré comme true");
        hornet_test_assert_1.HornetTestAssert.assertTrue("titi" == "titi", "titi == titi n'est pas considéré comme true");
        hornet_test_assert_1.HornetTestAssert.assertTrue(obj == obj, "objet == objet n'est pas considéré comme true");
        this.end();
    };
    ;
    HornetTestAssertKarma.prototype.assertFalseTest = function () {
        var obj1 = { label: "titi", value: 2 };
        var obj2 = { label: "toto", value: 5 };
        var value1 = 1;
        var value2 = 2;
        var value3 = "titi";
        var value4 = "toto";
        hornet_test_assert_1.HornetTestAssert.assertFalse(false, "false n'est pas considéré comme false");
        hornet_test_assert_1.HornetTestAssert.assertFalse(value1 == value2, "1 == 2 n'est pas considéré comme false");
        hornet_test_assert_1.HornetTestAssert.assertFalse(value3 == value4, "titi == toto n'est pas considéré comme false");
        hornet_test_assert_1.HornetTestAssert.assertFalse(obj1 == obj2, "obj1 == obj2 n'est pas considéré comme false");
        this.end();
    };
    ;
    HornetTestAssertKarma.prototype.assertGreaterThanTest = function () {
        hornet_test_assert_1.HornetTestAssert.assertGreaterThan(1, 2, "2 n'est pas plus grand que 1");
        this.end();
    };
    ;
    HornetTestAssertKarma.prototype.assertLesserThanTest = function () {
        hornet_test_assert_1.HornetTestAssert.assertLesserThan(2, 1, "2 n'est pas plus grand que 1");
        this.end();
    };
    ;
    tslib_1.__decorate([
        decorators_1.Decorators.it("Tests pour assertEquals"),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", []),
        tslib_1.__metadata("design:returntype", void 0)
    ], HornetTestAssertKarma.prototype, "assertEqualsTest", null);
    tslib_1.__decorate([
        decorators_1.Decorators.it("Tests pour assertNotEquals"),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", []),
        tslib_1.__metadata("design:returntype", void 0)
    ], HornetTestAssertKarma.prototype, "assertNotEqualsTest", null);
    tslib_1.__decorate([
        decorators_1.Decorators.it("Tests pour assertNull"),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", []),
        tslib_1.__metadata("design:returntype", void 0)
    ], HornetTestAssertKarma.prototype, "assertNullTest", null);
    tslib_1.__decorate([
        decorators_1.Decorators.it("Tests pour assertNotNull"),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", []),
        tslib_1.__metadata("design:returntype", void 0)
    ], HornetTestAssertKarma.prototype, "assertNotNullTest", null);
    tslib_1.__decorate([
        decorators_1.Decorators.it("Tests pour assertTrue"),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", []),
        tslib_1.__metadata("design:returntype", void 0)
    ], HornetTestAssertKarma.prototype, "assertTrueTest", null);
    tslib_1.__decorate([
        decorators_1.Decorators.it("Tests pour assertFalse"),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", []),
        tslib_1.__metadata("design:returntype", void 0)
    ], HornetTestAssertKarma.prototype, "assertFalseTest", null);
    tslib_1.__decorate([
        decorators_1.Decorators.it("Tests pour assertGreaterThan"),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", []),
        tslib_1.__metadata("design:returntype", void 0)
    ], HornetTestAssertKarma.prototype, "assertGreaterThanTest", null);
    tslib_1.__decorate([
        decorators_1.Decorators.it("Tests pour assertLesserThan"),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", []),
        tslib_1.__metadata("design:returntype", void 0)
    ], HornetTestAssertKarma.prototype, "assertLesserThanTest", null);
    HornetTestAssertKarma = tslib_1.__decorate([
        decorators_1.Decorators.describe("Test Karma Assert")
    ], HornetTestAssertKarma);
    return HornetTestAssertKarma;
}(base_test_1.BaseTest));
//lancement des Tests
test_run_1.runTest(new HornetTestAssertKarma());
