// app/hornet-js-react-components/src/widget/table/action-button.js
"use strict";
/**
 * Copyright ou © ou Copr. Ministère de l'Europe et des Affaires étrangères (2017)
 * <p/>
 * pole-architecture.dga-dsi-psi@diplomatie.gouv.fr
 * <p/>
 * Ce logiciel est un programme informatique servant à faciliter la création
 * d'applications Web conformément aux référentiels généraux français : RGI, RGS et RGAA
 * <p/>
 * Ce logiciel est régi par la licence CeCILL soumise au droit français et
 * respectant les principes de diffusion des logiciels libres. Vous pouvez
 * utiliser, modifier et/ou redistribuer ce programme sous les conditions
 * de la licence CeCILL telle que diffusée par le CEA, le CNRS et l'INRIA
 * sur le site "http://www.cecill.info".
 * <p/>
 * En contrepartie de l'accessibilité au code source et des droits de copie,
 * de modification et de redistribution accordés par cette licence, il n'est
 * offert aux utilisateurs qu'une garantie limitée.  Pour les mêmes raisons,
 * seule une responsabilité restreinte pèse sur l'auteur du programme,  le
 * titulaire des droits patrimoniaux et les concédants successifs.
 * <p/>
 * A cet égard  l'attention de l'utilisateur est attirée sur les risques
 * associés au chargement,  à l'utilisation,  à la modification et/ou au
 * développement et à la reproduction du logiciel par l'utilisateur étant
 * donné sa spécificité de logiciel libre, qui peut le rendre complexe à
 * manipuler et qui le réserve donc à des développeurs et des professionnels
 * avertis possédant  des  connaissances  informatiques approfondies.  Les
 * utilisateurs sont donc invités à charger  et  tester  l'adéquation  du
 * logiciel à leurs besoins dans des conditions permettant d'assurer la
 * sécurité de leurs systèmes et ou de leurs données et, plus généralement,
 * à l'utiliser et l'exploiter dans les mêmes conditions de sécurité.
 * <p/>
 * Le fait que vous puissiez accéder à cet en-tête signifie que vous avez
 * pris connaissance de la licence CeCILL, et que vous en avez accepté les
 * termes.
 * <p/>
 * <p/>
 * Copyright or © or Copr. Ministry for Europe and Foreign Affairs (2017)
 * <p/>
 * pole-architecture.dga-dsi-psi@diplomatie.gouv.fr
 * <p/>
 * This software is a computer program whose purpose is to facilitate creation of
 * web application in accordance with french general repositories : RGI, RGS and RGAA.
 * <p/>
 * This software is governed by the CeCILL license under French law and
 * abiding by the rules of distribution of free software.  You can  use,
 * modify and/ or redistribute the software under the terms of the CeCILL
 * license as circulated by CEA, CNRS and INRIA at the following URL
 * "http://www.cecill.info".
 * <p/>
 * As a counterpart to the access to the source code and  rights to copy,
 * modify and redistribute granted by the license, users are provided only
 * with a limited warranty  and the software's author,  the holder of the
 * economic rights,  and the successive licensors  have only  limited
 * liability.
 * <p/>
 * In this respect, the user's attention is drawn to the risks associated
 * with loading,  using,  modifying and/or developing or reproducing the
 * software by the user in light of its specific status of free software,
 * that may mean  that it is complicated to manipulate,  and  that  also
 * therefore means  that it is reserved for developers  and  experienced
 * professionals having in-depth computer knowledge. Users are therefore
 * encouraged to load and test the software's suitability as regards their
 * requirements in conditions enabling the security of their systems and/or
 * data to be ensured and,  more generally, to use and operate it in the
 * same conditions as regards security.
 * <p/>
 * The fact that you are presently reading this means that you have had
 * knowledge of the CeCILL license and that you accept its terms.
 *
 */
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
/**
 * hornet-js-react-components - Ensemble des composants web React de base de hornet-js
 *
 * @author MEAE - Ministère de l'Europe et des Affaires étrangères
 * @version v5.1.1
 * @link git+https://github.com/diplomatiegouvfr/hornet-js.git
 * @license CECILL-2.1
 */
var hornet_js_utils_1 = require("hornet-js-utils");
var React = require("react");
var button_1 = require("hornet-js-react-components/src/widget/button/button");
var classNames = require("classnames");
var _ = require("lodash");
var logger = hornet_js_utils_1.Utils.getLogger("hornet-js-react-components.widget.table.action-button");
/**
 * Enumeration des types d'action
 */
var TypeAction;
(function (TypeAction) {
    TypeAction[TypeAction["ACTION_MASSE"] = 1] = "ACTION_MASSE";
    TypeAction[TypeAction["ACTION_UNITAIRE"] = 2] = "ACTION_UNITAIRE";
})(TypeAction = exports.TypeAction || (exports.TypeAction = {}));
var ActionButton = /** @class */ (function (_super) {
    tslib_1.__extends(ActionButton, _super);
    function ActionButton(props, context) {
        var _this = _super.call(this, props, context) || this;
        if (props.url) {
            _this.state.url = _this.genUrlWithParams(props.url, props.value);
        }
        _this.state.visible = true;
        if (props.visible) {
            _this.state.visible = props.visible(_this.props.value);
        }
        return _this;
    }
    /**
     * @inheritDoc
     */
    ActionButton.prototype.render = function () {
        var _this = this;
        logger.trace("render actionButtons");
        var classes = {};
        if (this.props.className) {
            classes[this.props.className] = true;
        }
        classes["picto-svg"] = true;
        classes["button-action"] = true;
        var img = null;
        if (typeof this.props.srcImg === "string") {
            img = React.createElement("img", { src: this.props.srcImg, className: this.props.classNameImg, alt: this.props.title });
        }
        else {
            img = this.props.srcImg;
        }
        var keyDownFunction = null;
        if (this.props.onKeyDown) {
            keyDownFunction = function (e) {
                _this.props.onKeyDown(e, _this.props.onClick || _this.onClick);
            };
        }
        return (this.state.visible ?
            React.createElement("a", { href: this.props.url || "#", className: classNames(classes), title: this.props.title, onClick: this.props.onClick || this.onClick, onKeyDown: keyDownFunction, "aria-haspopup": this.props.hasPopUp },
                img,
                React.createElement("span", { className: "hidden-label" }, this.props.title),
                React.createElement("span", { className: "label-button-action" }, this.props.label))
            : null);
    };
    /**
     * Click sur le lien
     */
    ActionButton.prototype.onClick = function (e) {
        if (this.props.messageAlert) {
            e.stopPropagation();
            this.props.showAlert(this.props.messageAlert, this.props.titleAlert, this.onAction);
        }
        else {
            this.onAction();
        }
    };
    /**
     * action sur la confirmation
     */
    ActionButton.prototype.onAction = function () {
        if (this.state.url) {
            window.location.href = this.state.url;
        }
        else if (this.props.action) {
            this.props.action(this.props.value, this.props.selectedItems);
        }
    };
    ActionButton.defaultProps = _.assign(button_1.Button.defaultProps, {
        displayedWithoutResult: false
    });
    return ActionButton;
}(button_1.Button));
exports.ActionButton = ActionButton;
