// app/hornet-js-react-components/test/widget/notification/notification-pop-in-form-to-notification-karma.js
"use strict";
/**
 * Copyright ou © ou Copr. Ministère de l'Europe et des Affaires étrangères (2017)
 * <p/>
 * pole-architecture.dga-dsi-psi@diplomatie.gouv.fr
 * <p/>
 * Ce logiciel est un programme informatique servant à faciliter la création
 * d'applications Web conformément aux référentiels généraux français : RGI, RGS et RGAA
 * <p/>
 * Ce logiciel est régi par la licence CeCILL soumise au droit français et
 * respectant les principes de diffusion des logiciels libres. Vous pouvez
 * utiliser, modifier et/ou redistribuer ce programme sous les conditions
 * de la licence CeCILL telle que diffusée par le CEA, le CNRS et l'INRIA
 * sur le site "http://www.cecill.info".
 * <p/>
 * En contrepartie de l'accessibilité au code source et des droits de copie,
 * de modification et de redistribution accordés par cette licence, il n'est
 * offert aux utilisateurs qu'une garantie limitée.  Pour les mêmes raisons,
 * seule une responsabilité restreinte pèse sur l'auteur du programme,  le
 * titulaire des droits patrimoniaux et les concédants successifs.
 * <p/>
 * A cet égard  l'attention de l'utilisateur est attirée sur les risques
 * associés au chargement,  à l'utilisation,  à la modification et/ou au
 * développement et à la reproduction du logiciel par l'utilisateur étant
 * donné sa spécificité de logiciel libre, qui peut le rendre complexe à
 * manipuler et qui le réserve donc à des développeurs et des professionnels
 * avertis possédant  des  connaissances  informatiques approfondies.  Les
 * utilisateurs sont donc invités à charger  et  tester  l'adéquation  du
 * logiciel à leurs besoins dans des conditions permettant d'assurer la
 * sécurité de leurs systèmes et ou de leurs données et, plus généralement,
 * à l'utiliser et l'exploiter dans les mêmes conditions de sécurité.
 * <p/>
 * Le fait que vous puissiez accéder à cet en-tête signifie que vous avez
 * pris connaissance de la licence CeCILL, et que vous en avez accepté les
 * termes.
 * <p/>
 * <p/>
 * Copyright or © or Copr. Ministry for Europe and Foreign Affairs (2017)
 * <p/>
 * pole-architecture.dga-dsi-psi@diplomatie.gouv.fr
 * <p/>
 * This software is a computer program whose purpose is to facilitate creation of
 * web application in accordance with french general repositories : RGI, RGS and RGAA.
 * <p/>
 * This software is governed by the CeCILL license under French law and
 * abiding by the rules of distribution of free software.  You can  use,
 * modify and/ or redistribute the software under the terms of the CeCILL
 * license as circulated by CEA, CNRS and INRIA at the following URL
 * "http://www.cecill.info".
 * <p/>
 * As a counterpart to the access to the source code and  rights to copy,
 * modify and redistribute granted by the license, users are provided only
 * with a limited warranty  and the software's author,  the holder of the
 * economic rights,  and the successive licensors  have only  limited
 * liability.
 * <p/>
 * In this respect, the user's attention is drawn to the risks associated
 * with loading,  using,  modifying and/or developing or reproducing the
 * software by the user in light of its specific status of free software,
 * that may mean  that it is complicated to manipulate,  and  that  also
 * therefore means  that it is reserved for developers  and  experienced
 * professionals having in-depth computer knowledge. Users are therefore
 * encouraged to load and test the software's suitability as regards their
 * requirements in conditions enabling the security of their systems and/or
 * data to be ensured and,  more generally, to use and operate it in the
 * same conditions as regards security.
 * <p/>
 * The fact that you are presently reading this means that you have had
 * knowledge of the CeCILL license and that you accept its terms.
 *
 */
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
/**
 * hornet-js-react-components - Ensemble des composants web React de base de hornet-js
 *
 * @author MEAE - Ministère de l'Europe et des Affaires étrangères
 * @version v5.1.1
 * @link git+https://github.com/diplomatiegouvfr/hornet-js.git
 * @license CECILL-2.1
 */
var React = require("react");
var assert = require("assert");
var hornet_react_test_1 = require("hornet-js-test/src/hornet-react-test");
var hornet_test_assert_1 = require("hornet-js-test/src/hornet-test-assert");
var test_run_1 = require("hornet-js-test/src/test-run");
var decorators_1 = require("hornet-js-test/src/decorators");
var form_1 = require("hornet-js-react-components/src/widget/form/form");
var buttons_area_1 = require("hornet-js-react-components/src/widget/form/buttons-area");
var button_1 = require("hornet-js-react-components/src/widget/button/button");
var notification_1 = require("hornet-js-react-components/src/widget/notification/notification");
var notification_manager_1 = require("hornet-js-core/src/notification/notification-manager");
var modal_1 = require("hornet-js-react-components/src/widget/dialog/modal");
var input_field_1 = require("hornet-js-react-components/src/widget/form/input-field");
var schema2 = require("test/widget/notification/validation2.json");
var formElement;
var renderedElement;
var maModal1;
var maModal2;
/**
 * Tests visant à s'assurer que l'on peut dispatcher les notifications de formulaire
 * d'une pop in vers une zone de notification sur une page
 */
var NotificationFormDispatchingTest = /** @class */ (function (_super) {
    tslib_1.__extends(NotificationFormDispatchingTest, _super);
    function NotificationFormDispatchingTest() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    NotificationFormDispatchingTest.prototype.before = function () {
        formElement = (React.createElement("div", null,
            React.createElement(notification_1.Notification, { id: "notifZone" }),
            React.createElement(modal_1.Modal, { ref: function (modal) {
                    maModal1 = modal;
                }, withoutOverflow: true, underlayClickExits: false, focusDialog: false, onClickClose: this.closeModal1 },
                React.createElement(form_1.Form, { id: "modalForm1", onSubmit: this.onSubmitModalForm1 }, this.renderButtons("modalForm1"))),
            React.createElement(modal_1.Modal, { ref: function (modal) {
                    maModal2 = modal;
                }, withoutOverflow: true, underlayClickExits: false, focusDialog: false, onClickClose: this.closeModal2 },
                React.createElement(form_1.Form, { id: "modalForm2", onSubmit: this.onSubmitModalForm2, schema: schema2 },
                    React.createElement(input_field_1.InputField, { name: "input6", label: "Champ à renseigner", required: true, maxLength: 250 }),
                    this.renderButtons("modalForm2")))));
        renderedElement = this.renderIntoDocument(formElement, "main1");
        maModal1.open();
    };
    ;
    NotificationFormDispatchingTest.prototype.closeModal1 = function () {
        maModal1.close();
    };
    NotificationFormDispatchingTest.prototype.closeModal2 = function () {
        maModal2.close();
    };
    NotificationFormDispatchingTest.prototype.onSubmitModalForm1 = function () {
        maModal1.close();
        notification_manager_1.NotificationManager.cleanAll();
        notification_manager_1.NotificationManager.notify(null, null, null, notification_manager_1.Notifications.makeSingleNotification("", "Formulaire 1 Validé"));
    };
    NotificationFormDispatchingTest.prototype.onSubmitModalForm2 = function () {
        maModal2.close();
        notification_manager_1.NotificationManager.cleanAll();
        notification_manager_1.NotificationManager.notify(null, null, null, notification_manager_1.Notifications.makeSingleNotification("", "Formulaire 2 Validé"));
    };
    NotificationFormDispatchingTest.prototype.renderButtons = function (formName) {
        return (React.createElement(buttons_area_1.ButtonsArea, null,
            React.createElement(button_1.Button, { type: "submit", id: "envoi-" + formName, name: "action:" + formName, className: "hornet-button", value: "valider", label: "valider", title: "valider" })));
    };
    NotificationFormDispatchingTest.prototype.onSubmit = function () {
        var notifText = "Validation du formulaire";
        console.log(notifText);
    };
    NotificationFormDispatchingTest.prototype.testOk = function () {
        assert.equal(1, 1);
        this.end();
    };
    ;
    /**
     * Le formulaire 1 envoie ses notifications info vers la zone de notification partagée
     */
    NotificationFormDispatchingTest.prototype.validerPopInForm = function () {
        var _this = this;
        this.triggerMouseEvent(document.querySelector("#envoi-modalForm1"), "click");
        setTimeout(function () {
            hornet_test_assert_1.HornetTestAssert.assertFalse(maModal1.state.isVisible, "le modal modal1 ne doit plus être visible");
            var element = _this.getNotificationMessageListForm("notifZone", "info-message-list");
            hornet_test_assert_1.HornetTestAssert.assertEquals(1, element.length, "La zone de notification doit contenir un message d'info");
            _this.end();
        }, 500);
    };
    ;
    /**
     * Le formulaire 2 envoie ses notifications d'erreur dans la pop in
     */
    NotificationFormDispatchingTest.prototype.validerKOPopInForm = function () {
        var _this = this;
        maModal2.open();
        setTimeout(function () {
            _this.triggerMouseEvent(document.querySelector("#envoi-modalForm2"), "click");
            setTimeout(function () {
                hornet_test_assert_1.HornetTestAssert.assertTrue(maModal2.state.isVisible, "Le modal maModal2 doit être visible");
                var element = _this.getNotificationMessageListForm("Form-1", "error-message-list");
                hornet_test_assert_1.HornetTestAssert.assertEquals(1, element.length, "La zone de notification doit contenir un message d'error");
                _this.end();
            }, 500);
        }, 500);
    };
    ;
    NotificationFormDispatchingTest.prototype.getNotificationMessageListForm = function (form, className) {
        var formElement = document.getElementById(form);
        var messageList = formElement.getElementsByClassName(className)[0];
        return (messageList) ? messageList.children : null;
    };
    tslib_1.__decorate([
        decorators_1.Decorators.before,
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", []),
        tslib_1.__metadata("design:returntype", void 0)
    ], NotificationFormDispatchingTest.prototype, "before", null);
    tslib_1.__decorate([
        decorators_1.Decorators.it("Test OK"),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", []),
        tslib_1.__metadata("design:returntype", void 0)
    ], NotificationFormDispatchingTest.prototype, "testOk", null);
    tslib_1.__decorate([
        decorators_1.Decorators.it("Validation ok du formulaire d'une pop in"),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", []),
        tslib_1.__metadata("design:returntype", void 0)
    ], NotificationFormDispatchingTest.prototype, "validerPopInForm", null);
    tslib_1.__decorate([
        decorators_1.Decorators.it("Validation ko du formulaire d'une pop in"),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", []),
        tslib_1.__metadata("design:returntype", void 0)
    ], NotificationFormDispatchingTest.prototype, "validerKOPopInForm", null);
    NotificationFormDispatchingTest = tslib_1.__decorate([
        decorators_1.Decorators.describe("Test Karma Notification pop in to page")
    ], NotificationFormDispatchingTest);
    return NotificationFormDispatchingTest;
}(hornet_react_test_1.HornetReactTest));
//lancement des Tests
test_run_1.runTest(new NotificationFormDispatchingTest());
