// app/hornet-js-core/src/notification/notification-manager.js
"use strict";
/**
 * Copyright ou © ou Copr. Ministère de l'Europe et des Affaires étrangères (2017)
 * <p/>
 * pole-architecture.dga-dsi-psi@diplomatie.gouv.fr
 * <p/>
 * Ce logiciel est un programme informatique servant à faciliter la création
 * d'applications Web conformément aux référentiels généraux français : RGI, RGS et RGAA
 * <p/>
 * Ce logiciel est régi par la licence CeCILL soumise au droit français et
 * respectant les principes de diffusion des logiciels libres. Vous pouvez
 * utiliser, modifier et/ou redistribuer ce programme sous les conditions
 * de la licence CeCILL telle que diffusée par le CEA, le CNRS et l'INRIA
 * sur le site "http://www.cecill.info".
 * <p/>
 * En contrepartie de l'accessibilité au code source et des droits de copie,
 * de modification et de redistribution accordés par cette licence, il n'est
 * offert aux utilisateurs qu'une garantie limitée.  Pour les mêmes raisons,
 * seule une responsabilité restreinte pèse sur l'auteur du programme,  le
 * titulaire des droits patrimoniaux et les concédants successifs.
 * <p/>
 * A cet égard  l'attention de l'utilisateur est attirée sur les risques
 * associés au chargement,  à l'utilisation,  à la modification et/ou au
 * développement et à la reproduction du logiciel par l'utilisateur étant
 * donné sa spécificité de logiciel libre, qui peut le rendre complexe à
 * manipuler et qui le réserve donc à des développeurs et des professionnels
 * avertis possédant  des  connaissances  informatiques approfondies.  Les
 * utilisateurs sont donc invités à charger  et  tester  l'adéquation  du
 * logiciel à leurs besoins dans des conditions permettant d'assurer la
 * sécurité de leurs systèmes et ou de leurs données et, plus généralement,
 * à l'utiliser et l'exploiter dans les mêmes conditions de sécurité.
 * <p/>
 * Le fait que vous puissiez accéder à cet en-tête signifie que vous avez
 * pris connaissance de la licence CeCILL, et que vous en avez accepté les
 * termes.
 * <p/>
 * <p/>
 * Copyright or © or Copr. Ministry for Europe and Foreign Affairs (2017)
 * <p/>
 * pole-architecture.dga-dsi-psi@diplomatie.gouv.fr
 * <p/>
 * This software is a computer program whose purpose is to facilitate creation of
 * web application in accordance with french general repositories : RGI, RGS and RGAA.
 * <p/>
 * This software is governed by the CeCILL license under French law and
 * abiding by the rules of distribution of free software.  You can  use,
 * modify and/ or redistribute the software under the terms of the CeCILL
 * license as circulated by CEA, CNRS and INRIA at the following URL
 * "http://www.cecill.info".
 * <p/>
 * As a counterpart to the access to the source code and  rights to copy,
 * modify and redistribute granted by the license, users are provided only
 * with a limited warranty  and the software's author,  the holder of the
 * economic rights,  and the successive licensors  have only  limited
 * liability.
 * <p/>
 * In this respect, the user's attention is drawn to the risks associated
 * with loading,  using,  modifying and/or developing or reproducing the
 * software by the user in light of its specific status of free software,
 * that may mean  that it is complicated to manipulate,  and  that  also
 * therefore means  that it is reserved for developers  and  experienced
 * professionals having in-depth computer knowledge. Users are therefore
 * encouraged to load and test the software's suitability as regards their
 * requirements in conditions enabling the security of their systems and/or
 * data to be ensured and,  more generally, to use and operate it in the
 * same conditions as regards security.
 * <p/>
 * The fact that you are presently reading this means that you have had
 * knowledge of the CeCILL license and that you accept its terms.
 *
 */
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * hornet-js-core - Ensemble des composants qui forment le coeur de hornet-js
 *
 * @author MEAE - Ministère de l'Europe et des Affaires étrangères
 * @version v5.1.1
 * @link git+https://github.com/diplomatiegouvfr/hornet-js.git
 * @license CECILL-2.1
 */
var notification_events_1 = require("hornet-js-core/src/notification/notification-events");
var hornet_event_1 = require("hornet-js-core/src/event/hornet-event");
/**
 * Gestionnaire d'évènements de notifications
 */
var NotificationManager = /** @class */ (function () {
    function NotificationManager() {
    }
    NotificationManager.clean = function (id, idComponent) {
        hornet_event_1.fireHornetEvent(notification_events_1.CLEAN_NOTIFICATION_EVENT.withData({ id: id, idComponent: idComponent }));
    };
    NotificationManager.cleanAll = function () {
        hornet_event_1.fireHornetEvent(notification_events_1.CLEAN_ALL_NOTIFICATION_EVENT);
    };
    /**
     * Déclenche un évènement d'ajout de notification contenant les détails indiqués
     * @param id identifiant de notification
     * @param idComponent identifiant du composant déclenchant la notification
     * @param errors détail des erreurs éventuelles
     * @param infos informations éventuelles détail des informations éventuelles
     * @param exceptions exceptions détail des exceptions éventuelles
     * @param warnings détail des warnings éventuelles
     */
    NotificationManager.notify = function (id, idComponent, errors, infos, exceptions, warnings, personnals) {
        hornet_event_1.fireHornetEvent(notification_events_1.ADD_NOTIFICATION_EVENT.withData({ id: id, idComponent: idComponent, errors: errors, infos: infos, exceptions: exceptions, warnings: warnings, personnals: personnals }));
    };
    return NotificationManager;
}());
exports.NotificationManager = NotificationManager;
var Notifications = /** @class */ (function () {
    function Notifications(color, logo) {
        this.notifications = new Array();
        this.canRenderRealComponent = false;
        this.color = (color) ? color : "black";
        this.logo = (logo) ? logo : "";
    }
    Notifications.prototype.getNotifications = function () {
        return this.notifications;
    };
    Notifications.prototype.setNotifications = function (notifs) {
        this.notifications = notifs;
    };
    Notifications.prototype.getCanRenderRealComponent = function () {
        return this.canRenderRealComponent;
    };
    Notifications.prototype.addNotification = function (notification) {
        this.notifications.push(notification);
    };
    Notifications.prototype.addNotifications = function (notifications) {
        this.notifications = this.notifications.concat(notifications);
    };
    /**
     * Construit une instance de Notifications contenant une seule notification ayant l'identifiant et le message indiqués
     * @param id identifiant de la notification à créer
     * @param text message de la notification
     */
    Notifications.makeSingleNotification = function (id, text) {
        var notif = new NotificationType();
        notif.id = id;
        notif.text = text;
        var notifs = new Notifications();
        notifs.addNotification(notif);
        return notifs;
    };
    return Notifications;
}());
exports.Notifications = Notifications;
var NotificationType = /** @class */ (function () {
    function NotificationType() {
        this.id = "";
        this.text = "";
        this.anchor = "";
        this.field = "";
        this.canRenderRealComponent = false;
    }
    NotificationType.prototype.toString = function () {
        return "id:" + this.id + ", text:" + this.text;
    };
    return NotificationType;
}());
exports.NotificationType = NotificationType;
