// app/hornet-js-react-components/test/widget/table/table.paginate.array.test.karma.js
/**
 * Copyright ou © ou Copr. Ministère de l'Europe et des Affaires étrangères (2017)
 * <p/>
 * pole-architecture.dga-dsi-psi@diplomatie.gouv.fr
 * <p/>
 * Ce logiciel est un programme informatique servant à faciliter la création
 * d'applications Web conformément aux référentiels généraux français : RGI, RGS et RGAA
 * <p/>
 * Ce logiciel est régi par la licence CeCILL soumise au droit français et
 * respectant les principes de diffusion des logiciels libres. Vous pouvez
 * utiliser, modifier et/ou redistribuer ce programme sous les conditions
 * de la licence CeCILL telle que diffusée par le CEA, le CNRS et l'INRIA
 * sur le site "http://www.cecill.info".
 * <p/>
 * En contrepartie de l'accessibilité au code source et des droits de copie,
 * de modification et de redistribution accordés par cette licence, il n'est
 * offert aux utilisateurs qu'une garantie limitée.  Pour les mêmes raisons,
 * seule une responsabilité restreinte pèse sur l'auteur du programme,  le
 * titulaire des droits patrimoniaux et les concédants successifs.
 * <p/>
 * A cet égard  l'attention de l'utilisateur est attirée sur les risques
 * associés au chargement,  à l'utilisation,  à la modification et/ou au
 * développement et à la reproduction du logiciel par l'utilisateur étant
 * donné sa spécificité de logiciel libre, qui peut le rendre complexe à
 * manipuler et qui le réserve donc à des développeurs et des professionnels
 * avertis possédant  des  connaissances  informatiques approfondies.  Les
 * utilisateurs sont donc invités à charger  et  tester  l'adéquation  du
 * logiciel à leurs besoins dans des conditions permettant d'assurer la
 * sécurité de leurs systèmes et ou de leurs données et, plus généralement,
 * à l'utiliser et l'exploiter dans les mêmes conditions de sécurité.
 * <p/>
 * Le fait que vous puissiez accéder à cet en-tête signifie que vous avez
 * pris connaissance de la licence CeCILL, et que vous en avez accepté les
 * termes.
 * <p/>
 * <p/>
 * Copyright or © or Copr. Ministry for Europe and Foreign Affairs (2017)
 * <p/>
 * pole-architecture.dga-dsi-psi@diplomatie.gouv.fr
 * <p/>
 * This software is a computer program whose purpose is to facilitate creation of
 * web application in accordance with french general repositories : RGI, RGS and RGAA.
 * <p/>
 * This software is governed by the CeCILL license under French law and
 * abiding by the rules of distribution of free software.  You can  use,
 * modify and/ or redistribute the software under the terms of the CeCILL
 * license as circulated by CEA, CNRS and INRIA at the following URL
 * "http://www.cecill.info".
 * <p/>
 * As a counterpart to the access to the source code and  rights to copy,
 * modify and redistribute granted by the license, users are provided only
 * with a limited warranty  and the software's author,  the holder of the
 * economic rights,  and the successive licensors  have only  limited
 * liability.
 * <p/>
 * In this respect, the user's attention is drawn to the risks associated
 * with loading,  using,  modifying and/or developing or reproducing the
 * software by the user in light of its specific status of free software,
 * that may mean  that it is complicated to manipulate,  and  that  also
 * therefore means  that it is reserved for developers  and  experienced
 * professionals having in-depth computer knowledge. Users are therefore
 * encouraged to load and test the software's suitability as regards their
 * requirements in conditions enabling the security of their systems and/or
 * data to be ensured and,  more generally, to use and operate it in the
 * same conditions as regards security.
 * <p/>
 * The fact that you are presently reading this means that you have had
 * knowledge of the CeCILL license and that you accept its terms.
 *
 */
/**
 * hornet-js-react-components - Ensemble des composants web React de base de hornet-js
 *
 * @author MEAE - Ministère de l'Europe et des Affaires étrangères
 * @version v5.1.1
 * @link git+https://github.com/diplomatiegouvfr/hornet-js.git
 * @license CECILL-2.1
 */
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var chai = require('chai');
var expect = chai.expect;
var React = require("react");
var base_test_1 = require("hornet-js-test/src/base-test");
var test_run_1 = require("hornet-js-test/src/test-run");
var decorators_1 = require("hornet-js-test/src/decorators");
var assert = require("assert");
var paginate_datasource_1 = require("hornet-js-core/src/component/datasource/paginate-datasource");
var table_1 = require("hornet-js-react-components/src/widget/table/table");
var header_1 = require("hornet-js-react-components/src/widget/table/header");
/* Composant Content */
var content_1 = require("hornet-js-react-components/src/widget/table/content");
/*  Colonne du tableau */
var column_1 = require("hornet-js-react-components/src/widget/table/column");
var columns_1 = require("hornet-js-react-components/src/widget/table/columns");
var check_column_1 = require("hornet-js-react-components/src/widget/table/column/check-column");
var footer_1 = require("hornet-js-react-components/src/widget/table/footer");
var pager_1 = require("hornet-js-react-components/src/widget/pager/pager");
/** Tableau de liste de secteurs */
var dataSource;
var tableElement;
var table;
var data;
var tableTest = /** @class */ (function (_super) {
    tslib_1.__extends(tableTest, _super);
    function tableTest() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    tableTest.prototype.beforeEach = function () {
        data = [];
        var step = 1;
        for (var i = 1; i < 50; i++) {
            data.push({ id: i, label: "libelle" + i, desc: (step % 3 == 0) ? "desc" + 0 : "desc" + step++ });
        }
        dataSource = new paginate_datasource_1.PaginateDataSource(data, {
            pageIndex: 0,
            itemsPerPage: 10,
            totalItems: 0
        }, {});
        tableElement = (React.createElement(table_1.Table, { id: "lite" },
            React.createElement(header_1.Header, { title: "Secteurs" }),
            React.createElement(content_1.Content, { dataSource: dataSource },
                React.createElement(columns_1.Columns, null,
                    React.createElement(check_column_1.CheckColumn, { keyColumn: "id" }),
                    React.createElement(column_1.Column, { keyColumn: "label", title: "libelle", sortable: true }),
                    React.createElement(column_1.Column, { keyColumn: "desc", title: "desc", sortable: true }))),
            React.createElement(footer_1.Footer, null,
                React.createElement(pager_1.Pager, { dataSource: dataSource, id: "maTable-paginate" }))));
    };
    ;
    tableTest.prototype.testOk = function () {
        assert.equal(1, 1);
        this.end();
    };
    ;
    tableTest.prototype.selectionUnElement = function () {
        table = this.renderIntoDocument(tableElement, "main1");
        expect(document.querySelectorAll('#main1 .datatable-data tr').length).to.equal(10);
        this.end();
    };
    ;
    tableTest.prototype.goToLastPage = function () {
        var _this = this;
        table = this.renderIntoDocument(tableElement, "main2");
        dataSource.on("pagination", function (value) {
            expect(value.list.length).to.equal(9);
            expect(value.list[0].id).to.equal(41);
            expect(document.querySelector('#main2 .datatable-data #lite-0-colBody-0-1').innerHTML, value.list[0].label);
            _this.end();
        });
        this.triggerMouseEvent(document.querySelector('#main2 .datatable-pagination-button-lastpage'), "click");
    };
    ;
    tableTest.prototype.deleteSelectedItem = function () {
        var _this = this;
        table = this.renderIntoDocument(tableElement, "main3");
        dataSource.on("delete", function () {
            expect(dataSource.selected === undefined).to.true;
            _this.end();
        });
        dataSource.on("select", function (value) {
            if (value.length > 0)
                dataSource.deleteAll();
        });
        this.triggerMouseEvent(document.querySelector('#main3 .datatable-data #lite-0-colBody-0-0 input'), "click");
    };
    ;
    tableTest.prototype.goToPageCall = function () {
        var _this = this;
        table = this.renderIntoDocument(tableElement, "main4");
        dataSource.on("pagination", function (value) {
            expect(document.querySelector('#main4 .datatable-pagination-input')["value"]).to.equal("2");
            _this.end();
        });
        dataSource.goToPage(2);
    };
    ;
    tslib_1.__decorate([
        decorators_1.Decorators.beforeEach,
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", []),
        tslib_1.__metadata("design:returntype", void 0)
    ], tableTest.prototype, "beforeEach", null);
    tslib_1.__decorate([
        decorators_1.Decorators.it('Test OK'),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", []),
        tslib_1.__metadata("design:returntype", void 0)
    ], tableTest.prototype, "testOk", null);
    tslib_1.__decorate([
        decorators_1.Decorators.it('afficher 10 éléments par page à l\'init'),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", []),
        tslib_1.__metadata("design:returntype", void 0)
    ], tableTest.prototype, "selectionUnElement", null);
    tslib_1.__decorate([
        decorators_1.Decorators.it('afficher la dernière page'),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", []),
        tslib_1.__metadata("design:returntype", void 0)
    ], tableTest.prototype, "goToLastPage", null);
    tslib_1.__decorate([
        decorators_1.Decorators.it('supprimer la selection'),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", []),
        tslib_1.__metadata("design:returntype", void 0)
    ], tableTest.prototype, "deleteSelectedItem", null);
    tslib_1.__decorate([
        decorators_1.Decorators.it('test d\'un appel de pagination programmatiquement'),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", []),
        tslib_1.__metadata("design:returntype", void 0)
    ], tableTest.prototype, "goToPageCall", null);
    tableTest = tslib_1.__decorate([
        decorators_1.Decorators.describe('Test Karma table paginate')
    ], tableTest);
    return tableTest;
}(base_test_1.BaseTest));
//lancement des Tests
test_run_1.runTest(new tableTest());
