// app/hornet-js-components/src/event/key-codes.js
"use strict";
/**
 * Copyright ou © ou Copr. Ministère de l'Europe et des Affaires étrangères (2017)
 * <p/>
 * pole-architecture.dga-dsi-psi@diplomatie.gouv.fr
 * <p/>
 * Ce logiciel est un programme informatique servant à faciliter la création
 * d'applications Web conformément aux référentiels généraux français : RGI, RGS et RGAA
 * <p/>
 * Ce logiciel est régi par la licence CeCILL soumise au droit français et
 * respectant les principes de diffusion des logiciels libres. Vous pouvez
 * utiliser, modifier et/ou redistribuer ce programme sous les conditions
 * de la licence CeCILL telle que diffusée par le CEA, le CNRS et l'INRIA
 * sur le site "http://www.cecill.info".
 * <p/>
 * En contrepartie de l'accessibilité au code source et des droits de copie,
 * de modification et de redistribution accordés par cette licence, il n'est
 * offert aux utilisateurs qu'une garantie limitée.  Pour les mêmes raisons,
 * seule une responsabilité restreinte pèse sur l'auteur du programme,  le
 * titulaire des droits patrimoniaux et les concédants successifs.
 * <p/>
 * A cet égard  l'attention de l'utilisateur est attirée sur les risques
 * associés au chargement,  à l'utilisation,  à la modification et/ou au
 * développement et à la reproduction du logiciel par l'utilisateur étant
 * donné sa spécificité de logiciel libre, qui peut le rendre complexe à
 * manipuler et qui le réserve donc à des développeurs et des professionnels
 * avertis possédant  des  connaissances  informatiques approfondies.  Les
 * utilisateurs sont donc invités à charger  et  tester  l'adéquation  du
 * logiciel à leurs besoins dans des conditions permettant d'assurer la
 * sécurité de leurs systèmes et ou de leurs données et, plus généralement,
 * à l'utiliser et l'exploiter dans les mêmes conditions de sécurité.
 * <p/>
 * Le fait que vous puissiez accéder à cet en-tête signifie que vous avez
 * pris connaissance de la licence CeCILL, et que vous en avez accepté les
 * termes.
 * <p/>
 * <p/>
 * Copyright or © or Copr. Ministry for Europe and Foreign Affairs (2017)
 * <p/>
 * pole-architecture.dga-dsi-psi@diplomatie.gouv.fr
 * <p/>
 * This software is a computer program whose purpose is to facilitate creation of
 * web application in accordance with french general repositories : RGI, RGS and RGAA.
 * <p/>
 * This software is governed by the CeCILL license under French law and
 * abiding by the rules of distribution of free software.  You can  use,
 * modify and/ or redistribute the software under the terms of the CeCILL
 * license as circulated by CEA, CNRS and INRIA at the following URL
 * "http://www.cecill.info".
 * <p/>
 * As a counterpart to the access to the source code and  rights to copy,
 * modify and redistribute granted by the license, users are provided only
 * with a limited warranty  and the software's author,  the holder of the
 * economic rights,  and the successive licensors  have only  limited
 * liability.
 * <p/>
 * In this respect, the user's attention is drawn to the risks associated
 * with loading,  using,  modifying and/or developing or reproducing the
 * software by the user in light of its specific status of free software,
 * that may mean  that it is complicated to manipulate,  and  that  also
 * therefore means  that it is reserved for developers  and  experienced
 * professionals having in-depth computer knowledge. Users are therefore
 * encouraged to load and test the software's suitability as regards their
 * requirements in conditions enabling the security of their systems and/or
 * data to be ensured and,  more generally, to use and operate it in the
 * same conditions as regards security.
 * <p/>
 * The fact that you are presently reading this means that you have had
 * knowledge of the CeCILL license and that you accept its terms.
 *
 */
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * hornet-js-components - Interfaces des composants web de hornet-js
 *
 * @author MEAE - Ministère de l'Europe et des Affaires étrangères
 * @version v5.1.1
 * @link git+https://github.com/diplomatiegouvfr/hornet-js.git
 * @license CECILL-2.1
 */
/**
 * Enumération de valeurs pour DOM KeyboardEvent.keyCode.
 * Cette attribut est obsolète
 * (cf. https://www.w3.org/TR/DOM-Level-3-Events/#legacy-key-attributes)
 * mais est mieux supporté que event.code ou event.key par les différents navigateurs
 * et semble avoir des valeurs plus cohérentes entre navigateurs.
 * Différences de valeurs pour event.code : https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/code#Code_values)
 * Safari ne prend pas en charge event.code et event.key : https://bugs.webkit.org/show_bug.cgi?id=149584,
 * https://bugs.webkit.org/show_bug.cgi?id=69029
 * IE ne prend pas en charge event.code */
var KeyCodes;
(function (KeyCodes) {
    KeyCodes[KeyCodes["BREAK"] = 3] = "BREAK";
    KeyCodes[KeyCodes["BACKSPACE"] = 8] = "BACKSPACE";
    KeyCodes[KeyCodes["TAB"] = 9] = "TAB";
    KeyCodes[KeyCodes["CLEAR"] = 12] = "CLEAR";
    KeyCodes[KeyCodes["ENTER"] = 13] = "ENTER";
    KeyCodes[KeyCodes["SHIFT"] = 16] = "SHIFT";
    KeyCodes[KeyCodes["CTRL"] = 17] = "CTRL";
    KeyCodes[KeyCodes["ALT"] = 18] = "ALT";
    KeyCodes[KeyCodes["PAUSE"] = 19] = "PAUSE";
    KeyCodes[KeyCodes["CAPS_LOCK"] = 20] = "CAPS_LOCK";
    KeyCodes[KeyCodes["ESCAPE"] = 27] = "ESCAPE";
    KeyCodes[KeyCodes["SPACEBAR"] = 32] = "SPACEBAR";
    KeyCodes[KeyCodes["PAGE_UP"] = 33] = "PAGE_UP";
    KeyCodes[KeyCodes["PAGE_DOWN"] = 34] = "PAGE_DOWN";
    KeyCodes[KeyCodes["END"] = 35] = "END";
    KeyCodes[KeyCodes["HOME"] = 36] = "HOME";
    KeyCodes[KeyCodes["LEFT_ARROW"] = 37] = "LEFT_ARROW";
    KeyCodes[KeyCodes["UP_ARROW"] = 38] = "UP_ARROW";
    KeyCodes[KeyCodes["RIGHT_ARROW"] = 39] = "RIGHT_ARROW";
    KeyCodes[KeyCodes["DOWN_ARROW"] = 40] = "DOWN_ARROW";
    KeyCodes[KeyCodes["SELECT"] = 41] = "SELECT";
    KeyCodes[KeyCodes["PRINT"] = 42] = "PRINT";
    KeyCodes[KeyCodes["EXECUTE"] = 43] = "EXECUTE";
    KeyCodes[KeyCodes["PRINT_SCREEN"] = 44] = "PRINT_SCREEN";
    KeyCodes[KeyCodes["INSERT"] = 45] = "INSERT";
    KeyCodes[KeyCodes["DELETE"] = 46] = "DELETE";
    KeyCodes[KeyCodes["KEY_0"] = 48] = "KEY_0";
    KeyCodes[KeyCodes["KEY_6"] = 54] = "KEY_6";
    KeyCodes[KeyCodes["KEY_9"] = 57] = "KEY_9";
    KeyCodes[KeyCodes["FF_SLASH"] = 58] = "FF_SLASH";
    KeyCodes[KeyCodes["KEY_A"] = 65] = "KEY_A";
    KeyCodes[KeyCodes["KEY_C"] = 67] = "KEY_C";
    KeyCodes[KeyCodes["KEY_V"] = 86] = "KEY_V";
    KeyCodes[KeyCodes["KEY_X"] = 88] = "KEY_X";
    KeyCodes[KeyCodes["WINDOWS_KEY"] = 91] = "WINDOWS_KEY";
    KeyCodes[KeyCodes["WINDOWS_KEY_RIGHT"] = 92] = "WINDOWS_KEY_RIGHT";
    KeyCodes[KeyCodes["WINDOWS_MENU"] = 93] = "WINDOWS_MENU";
    KeyCodes[KeyCodes["NUMPAD_0"] = 96] = "NUMPAD_0";
    KeyCodes[KeyCodes["NUMPAD_1"] = 97] = "NUMPAD_1";
    KeyCodes[KeyCodes["NUMPAD_2"] = 98] = "NUMPAD_2";
    KeyCodes[KeyCodes["NUMPAD_3"] = 99] = "NUMPAD_3";
    KeyCodes[KeyCodes["NUMPAD_4"] = 100] = "NUMPAD_4";
    KeyCodes[KeyCodes["NUMPAD_5"] = 101] = "NUMPAD_5";
    KeyCodes[KeyCodes["NUMPAD_6"] = 102] = "NUMPAD_6";
    KeyCodes[KeyCodes["NUMPAD_7"] = 103] = "NUMPAD_7";
    KeyCodes[KeyCodes["NUMPAD_8"] = 104] = "NUMPAD_8";
    KeyCodes[KeyCodes["NUMPAD_9"] = 105] = "NUMPAD_9";
    KeyCodes[KeyCodes["NUMPAD_MULTIPLY"] = 106] = "NUMPAD_MULTIPLY";
    KeyCodes[KeyCodes["NUMPAD_ADD"] = 107] = "NUMPAD_ADD";
    KeyCodes[KeyCodes["NUMPAD_SUBSTRACT"] = 109] = "NUMPAD_SUBSTRACT";
    KeyCodes[KeyCodes["NUMPAD_DECIMAL_POINT"] = 110] = "NUMPAD_DECIMAL_POINT";
    KeyCodes[KeyCodes["NUMPAD_DIVIDE"] = 111] = "NUMPAD_DIVIDE";
    KeyCodes[KeyCodes["F1"] = 112] = "F1";
    KeyCodes[KeyCodes["F2"] = 113] = "F2";
    KeyCodes[KeyCodes["F3"] = 114] = "F3";
    KeyCodes[KeyCodes["F4"] = 115] = "F4";
    KeyCodes[KeyCodes["F5"] = 116] = "F5";
    KeyCodes[KeyCodes["F6"] = 117] = "F6";
    KeyCodes[KeyCodes["F7"] = 118] = "F7";
    KeyCodes[KeyCodes["F8"] = 119] = "F8";
    KeyCodes[KeyCodes["F9"] = 120] = "F9";
    KeyCodes[KeyCodes["F10"] = 121] = "F10";
    KeyCodes[KeyCodes["F11"] = 122] = "F11";
    KeyCodes[KeyCodes["F12"] = 123] = "F12";
    KeyCodes[KeyCodes["F13"] = 124] = "F13";
    KeyCodes[KeyCodes["F14"] = 125] = "F14";
    KeyCodes[KeyCodes["F15"] = 126] = "F15";
    KeyCodes[KeyCodes["F16"] = 127] = "F16";
    KeyCodes[KeyCodes["F17"] = 128] = "F17";
    KeyCodes[KeyCodes["F18"] = 129] = "F18";
    KeyCodes[KeyCodes["F19"] = 130] = "F19";
    KeyCodes[KeyCodes["F20"] = 131] = "F20";
    KeyCodes[KeyCodes["F21"] = 132] = "F21";
    KeyCodes[KeyCodes["F22"] = 133] = "F22";
    KeyCodes[KeyCodes["F23"] = 134] = "F23";
    KeyCodes[KeyCodes["F24"] = 135] = "F24";
    KeyCodes[KeyCodes["NUM_LOCK"] = 144] = "NUM_LOCK";
    KeyCodes[KeyCodes["SCROLL_LOCK"] = 145] = "SCROLL_LOCK";
    KeyCodes[KeyCodes["POINT"] = 190] = "POINT";
    KeyCodes[KeyCodes["SLASH"] = 191] = "SLASH";
})(KeyCodes = exports.KeyCodes || (exports.KeyCodes = {}));
exports.KEYNAMES = {
    ArrowUp: "ArrowUp",
    ArrowDown: "ArrowDown",
    Home: "Home",
    Enter: "Enter"
};
