// app/hornet-js-react-components/src/widget/table/table-state.js
"use strict";
/**
 * Copyright ou © ou Copr. Ministère de l'Europe et des Affaires étrangères (2017)
 * <p/>
 * pole-architecture.dga-dsi-psi@diplomatie.gouv.fr
 * <p/>
 * Ce logiciel est un programme informatique servant à faciliter la création
 * d'applications Web conformément aux référentiels généraux français : RGI, RGS et RGAA
 * <p/>
 * Ce logiciel est régi par la licence CeCILL soumise au droit français et
 * respectant les principes de diffusion des logiciels libres. Vous pouvez
 * utiliser, modifier et/ou redistribuer ce programme sous les conditions
 * de la licence CeCILL telle que diffusée par le CEA, le CNRS et l'INRIA
 * sur le site "http://www.cecill.info".
 * <p/>
 * En contrepartie de l'accessibilité au code source et des droits de copie,
 * de modification et de redistribution accordés par cette licence, il n'est
 * offert aux utilisateurs qu'une garantie limitée.  Pour les mêmes raisons,
 * seule une responsabilité restreinte pèse sur l'auteur du programme,  le
 * titulaire des droits patrimoniaux et les concédants successifs.
 * <p/>
 * A cet égard  l'attention de l'utilisateur est attirée sur les risques
 * associés au chargement,  à l'utilisation,  à la modification et/ou au
 * développement et à la reproduction du logiciel par l'utilisateur étant
 * donné sa spécificité de logiciel libre, qui peut le rendre complexe à
 * manipuler et qui le réserve donc à des développeurs et des professionnels
 * avertis possédant  des  connaissances  informatiques approfondies.  Les
 * utilisateurs sont donc invités à charger  et  tester  l'adéquation  du
 * logiciel à leurs besoins dans des conditions permettant d'assurer la
 * sécurité de leurs systèmes et ou de leurs données et, plus généralement,
 * à l'utiliser et l'exploiter dans les mêmes conditions de sécurité.
 * <p/>
 * Le fait que vous puissiez accéder à cet en-tête signifie que vous avez
 * pris connaissance de la licence CeCILL, et que vous en avez accepté les
 * termes.
 * <p/>
 * <p/>
 * Copyright or © or Copr. Ministry for Europe and Foreign Affairs (2017)
 * <p/>
 * pole-architecture.dga-dsi-psi@diplomatie.gouv.fr
 * <p/>
 * This software is a computer program whose purpose is to facilitate creation of
 * web application in accordance with french general repositories : RGI, RGS and RGAA.
 * <p/>
 * This software is governed by the CeCILL license under French law and
 * abiding by the rules of distribution of free software.  You can  use,
 * modify and/ or redistribute the software under the terms of the CeCILL
 * license as circulated by CEA, CNRS and INRIA at the following URL
 * "http://www.cecill.info".
 * <p/>
 * As a counterpart to the access to the source code and  rights to copy,
 * modify and redistribute granted by the license, users are provided only
 * with a limited warranty  and the software's author,  the holder of the
 * economic rights,  and the successive licensors  have only  limited
 * liability.
 * <p/>
 * In this respect, the user's attention is drawn to the risks associated
 * with loading,  using,  modifying and/or developing or reproducing the
 * software by the user in light of its specific status of free software,
 * that may mean  that it is complicated to manipulate,  and  that  also
 * therefore means  that it is reserved for developers  and  experienced
 * professionals having in-depth computer knowledge. Users are therefore
 * encouraged to load and test the software's suitability as regards their
 * requirements in conditions enabling the security of their systems and/or
 * data to be ensured and,  more generally, to use and operate it in the
 * same conditions as regards security.
 * <p/>
 * The fact that you are presently reading this means that you have had
 * knowledge of the CeCILL license and that you accept its terms.
 *
 */
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
/**
 * hornet-js-react-components - Ensemble des composants web React de base de hornet-js
 *
 * @author MEAE - Ministère de l'Europe et des Affaires étrangères
 * @version v5.1.1
 * @link git+https://github.com/diplomatiegouvfr/hornet-js.git
 * @license CECILL-2.1
 */
var events = require("events");
var _ = require("lodash");
var TableState = /** @class */ (function (_super) {
    tslib_1.__extends(TableState, _super);
    function TableState() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.contentsState = [];
        return _this;
    }
    TableState.prototype.addContent = function (contentState) {
        contentState.on(ContentState.ITEMS_CHANGE_EVENT, this.setIndexAfterItemsChangeEvent.bind(this));
        this.contentsState.push(contentState);
    };
    TableState.prototype.removeContent = function (contentState) {
        contentState.removeListener(ContentState.ITEMS_CHANGE_EVENT, this.setIndexAfterItemsChangeEvent.bind(this));
        _.remove(this.contentsState, contentState);
    };
    TableState.prototype.removeContents = function () {
        var _this = this;
        this.contentsState.map(function (contentState) { return _this.removeContent(contentState); });
    };
    TableState.prototype.setIndexAfterItemsChangeEvent = function (items) {
        var _this = this;
        this.contentsState.map(function (contentState, index) {
            if (items === contentState.items) {
                _this.index = index;
                _this.emit(TableState.INDEX_CHANGE_EVENT, null, _this.contentsState[index].items);
            }
        });
    };
    TableState.prototype.getIndex = function () {
        return this.index;
    };
    TableState.INDEX_CHANGE_EVENT = "onIndexChange";
    TableState.RESIZE_EVENT = "onResizeChange";
    return TableState;
}(events.EventEmitter));
exports.TableState = TableState;
var ContentState = /** @class */ (function (_super) {
    tslib_1.__extends(ContentState, _super);
    function ContentState() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.items = [];
        _this.hiddenColumns = {};
        return _this;
    }
    ContentState.prototype.setFocusOn = function (focusedCell) {
        this.emit(ContentState.BLUR_EVENT, this.focusedCell);
        this.emit(ContentState.FOCUS_CHANGE_EVENT, this.focusedCell, focusedCell);
        this.focusedCell = focusedCell;
    };
    ContentState.prototype.setItemInEdition = function (itemInEditionValue, lineIndex) {
        this.itemInEdition = itemInEditionValue;
        if (this.itemInEdition) {
            this.itemInEdition["row"] = lineIndex;
        }
        this.emit(ContentState.EDITION_CLIC_EVENT, lineIndex);
    };
    ContentState.prototype.setItems = function (items) {
        this.items = items;
        this.emit(ContentState.ITEMS_CHANGE_EVENT, this.items);
    };
    ContentState.prototype.setHiddenColumns = function (hiddenColumns) {
        this.hiddenColumns = hiddenColumns;
        this.emit(ContentState.TOGGLE_COLUMNS_EVENT, this.hiddenColumns, this.firstVisibleColumnState, this.oldFirstVisibleColumnState);
    };
    ContentState.prototype.setFirstVisibleColumnState = function (columnState) {
        if (columnState != this.firstVisibleColumnState) {
            this.oldFirstVisibleColumnState = this.firstVisibleColumnState;
            this.firstVisibleColumnState = columnState;
        }
    };
    ContentState.prototype.setHasCheckColumnMassSelection = function (value) {
        this.hasCheckColumnMassSelection = value;
    };
    ContentState.prototype.setKeycolumnMassSelection = function (keyColumn) {
        this.keyColumnMassSelection = keyColumn;
    };
    ContentState.FOCUS_CHANGE_EVENT = "onFocusChange";
    ContentState.ITEMS_CHANGE_EVENT = "onItemsChange";
    ContentState.RESIZE_EVENT = "onResizeChange";
    ContentState.EDITION_CLIC_EVENT = "onEditionClic";
    ContentState.BLUR_EVENT = "onBlur";
    ContentState.TOGGLE_COLUMNS_EVENT = "toggleColumns";
    return ContentState;
}(events.EventEmitter));
exports.ContentState = ContentState;
