// app/hornet-js-react-components/test/widget/form/calendar-field.test.karma.js
/**
 * Copyright ou © ou Copr. Ministère de l'Europe et des Affaires étrangères (2017)
 * <p/>
 * pole-architecture.dga-dsi-psi@diplomatie.gouv.fr
 * <p/>
 * Ce logiciel est un programme informatique servant à faciliter la création
 * d'applications Web conformément aux référentiels généraux français : RGI, RGS et RGAA
 * <p/>
 * Ce logiciel est régi par la licence CeCILL soumise au droit français et
 * respectant les principes de diffusion des logiciels libres. Vous pouvez
 * utiliser, modifier et/ou redistribuer ce programme sous les conditions
 * de la licence CeCILL telle que diffusée par le CEA, le CNRS et l'INRIA
 * sur le site "http://www.cecill.info".
 * <p/>
 * En contrepartie de l'accessibilité au code source et des droits de copie,
 * de modification et de redistribution accordés par cette licence, il n'est
 * offert aux utilisateurs qu'une garantie limitée.  Pour les mêmes raisons,
 * seule une responsabilité restreinte pèse sur l'auteur du programme,  le
 * titulaire des droits patrimoniaux et les concédants successifs.
 * <p/>
 * A cet égard  l'attention de l'utilisateur est attirée sur les risques
 * associés au chargement,  à l'utilisation,  à la modification et/ou au
 * développement et à la reproduction du logiciel par l'utilisateur étant
 * donné sa spécificité de logiciel libre, qui peut le rendre complexe à
 * manipuler et qui le réserve donc à des développeurs et des professionnels
 * avertis possédant  des  connaissances  informatiques approfondies.  Les
 * utilisateurs sont donc invités à charger  et  tester  l'adéquation  du
 * logiciel à leurs besoins dans des conditions permettant d'assurer la
 * sécurité de leurs systèmes et ou de leurs données et, plus généralement,
 * à l'utiliser et l'exploiter dans les mêmes conditions de sécurité.
 * <p/>
 * Le fait que vous puissiez accéder à cet en-tête signifie que vous avez
 * pris connaissance de la licence CeCILL, et que vous en avez accepté les
 * termes.
 * <p/>
 * <p/>
 * Copyright or © or Copr. Ministry for Europe and Foreign Affairs (2017)
 * <p/>
 * pole-architecture.dga-dsi-psi@diplomatie.gouv.fr
 * <p/>
 * This software is a computer program whose purpose is to facilitate creation of
 * web application in accordance with french general repositories : RGI, RGS and RGAA.
 * <p/>
 * This software is governed by the CeCILL license under French law and
 * abiding by the rules of distribution of free software.  You can  use,
 * modify and/ or redistribute the software under the terms of the CeCILL
 * license as circulated by CEA, CNRS and INRIA at the following URL
 * "http://www.cecill.info".
 * <p/>
 * As a counterpart to the access to the source code and  rights to copy,
 * modify and redistribute granted by the license, users are provided only
 * with a limited warranty  and the software's author,  the holder of the
 * economic rights,  and the successive licensors  have only  limited
 * liability.
 * <p/>
 * In this respect, the user's attention is drawn to the risks associated
 * with loading,  using,  modifying and/or developing or reproducing the
 * software by the user in light of its specific status of free software,
 * that may mean  that it is complicated to manipulate,  and  that  also
 * therefore means  that it is reserved for developers  and  experienced
 * professionals having in-depth computer knowledge. Users are therefore
 * encouraged to load and test the software's suitability as regards their
 * requirements in conditions enabling the security of their systems and/or
 * data to be ensured and,  more generally, to use and operate it in the
 * same conditions as regards security.
 * <p/>
 * The fact that you are presently reading this means that you have had
 * knowledge of the CeCILL license and that you accept its terms.
 *
 */
/**
 * hornet-js-react-components - Ensemble des composants web React de base de hornet-js
 *
 * @author MEAE - Ministère de l'Europe et des Affaires étrangères
 * @version v5.1.1
 * @link git+https://github.com/diplomatiegouvfr/hornet-js.git
 * @license CECILL-2.1
 */
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var test_run_1 = require("hornet-js-test/src/test-run");
require("hornet-js-test/src/test-run");
var chai = require('chai');
var expect = chai.expect;
var React = require("react");
var hornet_react_test_1 = require("hornet-js-test/src/hornet-react-test");
var decorators_1 = require("hornet-js-test/src/decorators");
var assert = require("assert");
var hornet_js_utils_1 = require("hornet-js-utils");
var moment = require("moment");
var calendar_field_1 = require("hornet-js-react-components/src/widget/form/calendar-field");
var formElement;
var form;
var CalendarFieldTest = /** @class */ (function (_super) {
    tslib_1.__extends(CalendarFieldTest, _super);
    function CalendarFieldTest() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    CalendarFieldTest.prototype.beforeEach = function () {
        hornet_js_utils_1.Utils.setCls("hornet.internationalization", {
            messages: {
                "calendar": {
                    "agendaTitle": "Agenda",
                    "choiceDate": "Choisir une date",
                    "today": "Aujourd'hui",
                    "now": "Maintenant",
                    "ok": "Ok",
                    "clear": "Effacer",
                    "month": "Mois",
                    "year": "Année",
                    "hourPanelTitle": "Choisir une heure",
                    "minutePanelTitle": "Choisir une minute",
                    "secondPanelTitle": "Choisir une seconde",
                    "monthSelect": "Choisir un mois",
                    "yearSelect": "Choisir une année",
                    "decadeSelect": "Choisir une décade",
                    "yearFormat": "YYYY",
                    "dateFormat": "DD\u002FMM\u002FYYYY",
                    "monthYearFormat": "MMMM YYYY",
                    "monthFormat": "MMMM",
                    "monthBeforeYear": true,
                    "timezoneOffset": 60,
                    "firstDayOfWeek": 1,
                    "minimalDaysInFirstWeek": 1,
                    "placeHolder": "JJ\u002FMM\u002FAAAA"
                },
            }
        });
        formElement = (React.createElement(calendar_field_1.CalendarField, { name: "test", label: "Test date" }));
    };
    ;
    CalendarFieldTest.prototype.testOk = function () {
        assert.equal(1, 1);
        this.end();
    };
    ;
    CalendarFieldTest.prototype.ShowElement = function () {
        form = this.renderIntoDocument(formElement, "main1");
        expect(document.querySelector("#main1 .calendar-input")).to.exist;
        expect(document.querySelector("#main1 .agenda")).to.exist;
        this.end();
    };
    ;
    CalendarFieldTest.prototype.controlInputElement = function () {
        var _this = this;
        var SpecialCar = [
            { key: "²" }, { key: "<" }, { key: "&" }, { key: "é" }, { key: "~" }, { key: "#" },
            { key: "\"" }, { key: "{" }, { key: "\'" }, { key: "[" }, { key: "(" }, { key: "|" },
            { key: "`" }, { key: "è" }, { key: "_" }, { key: "\\" }, { key: "^" }, { key: "ç" },
            { key: "à" }, { key: "@" }, { key: ")" }, { key: "]" }, { key: "°" }, { key: "+" },
            { key: "=" }, { key: "}" }, { key: "£" }, { key: "$" }, { key: "¤" }, { key: "*" },
            { key: "µ" }, { key: "%" }, { key: "ù" }, { key: "!" }, { key: "§" }, { key: ":" },
            { key: "." }, { key: ";" }, { key: "?" }, { key: "," }, { key: "<" }, { key: ">" },
            { key: "æ" }, { key: "«" }, { key: "ð" }, { key: "¹" }, { key: "”" }, { key: "ß" }
        ];
        form = this.renderIntoDocument(formElement, "main2");
        var dateInput = (document.querySelector("#main2 .calendar-input"));
        setTimeout(function () {
            SpecialCar.forEach(function (char, index) {
                _this.triggerKeyPressEvent(dateInput, char.key, char.key.charCodeAt(0), false);
                expect(_this.isPreventDefault, "Impossible de saisir le caractère : " + char.key).be.true;
            });
            _this.end();
        }, 2500);
    };
    ;
    CalendarFieldTest.prototype.controlInvalidDateElement = function () {
        var _this = this;
        form = this.renderIntoDocument(React.createElement(calendar_field_1.CalendarField, { valideOnForm: false, label: "Test date", name: "dateTest" }), "main3");
        var dateInput = (document.querySelector("#main3 .calendar-input"));
        setTimeout(function () {
            _this.triggerKeyPressEvent(dateInput, "24/10/2012", 0, true);
            //this.triggerFocusOnElement(dateInput);
            _this.triggerFocusOnElement(document.querySelector("#main3 .agenda"));
            expect(document.querySelector("#main3 .fielderror-content")).to.not.exist;
            _this.end();
        }, 2500);
    };
    CalendarFieldTest.prototype.opentAndSelectedDate = function () {
        var _this = this;
        var onSelectCount = 0;
        var onChangeCount = 0;
        var onValueChangeCount = 0;
        form = this.renderIntoDocument(React.createElement(calendar_field_1.CalendarField, { valideOnForm: false, label: "Test date", name: "dateTest", onSelect: function (value) { onSelectCount++; }, onChange: function (value) { onChangeCount++; }, onValueChange: function (value) { onValueChangeCount++; } }), "main5");
        setTimeout(function () {
            _this.triggerMouseEvent(document.querySelector("#main5 .agenda"), "click");
            expect(document.querySelector(".widget-dialogue-header")).to.exist;
            _this.triggerMouseEvent(document.querySelector(".rc-calendar-selected-day"), "click");
            expect(document.querySelector("#main5 #dateTest").value).to.be.equal(moment().locale('fr').format("DD/MM/YYYY"));
            _this.triggerMouseEvent(document.querySelector("#main5 #dateTestResetButton a"), "click");
            expect(onSelectCount, "selectCount").to.be.equal(2);
            expect(onChangeCount, "changeCount").to.be.equal(1);
            expect(onValueChangeCount, "valueChangeCount").to.be.equal(2);
            _this.end();
        }, 3000);
    };
    tslib_1.__decorate([
        decorators_1.Decorators.beforeEach,
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", []),
        tslib_1.__metadata("design:returntype", void 0)
    ], CalendarFieldTest.prototype, "beforeEach", null);
    tslib_1.__decorate([
        decorators_1.Decorators.it('Test OK'),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", []),
        tslib_1.__metadata("design:returntype", void 0)
    ], CalendarFieldTest.prototype, "testOk", null);
    tslib_1.__decorate([
        decorators_1.Decorators.it('Affichage du champs calendar'),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", []),
        tslib_1.__metadata("design:returntype", void 0)
    ], CalendarFieldTest.prototype, "ShowElement", null);
    tslib_1.__decorate([
        decorators_1.Decorators.it('Test saisie de caractère non autorisé'),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", []),
        tslib_1.__metadata("design:returntype", void 0)
    ], CalendarFieldTest.prototype, "controlInputElement", null);
    tslib_1.__decorate([
        decorators_1.Decorators.it("Test saisie date valide"),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", []),
        tslib_1.__metadata("design:returntype", void 0)
    ], CalendarFieldTest.prototype, "controlInvalidDateElement", null);
    tslib_1.__decorate([
        decorators_1.Decorators.it("Test selection d'une date dans la boite de dialogue"),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", []),
        tslib_1.__metadata("design:returntype", void 0)
    ], CalendarFieldTest.prototype, "opentAndSelectedDate", null);
    CalendarFieldTest = tslib_1.__decorate([
        decorators_1.Decorators.describe('Test Karma calendar-field')
    ], CalendarFieldTest);
    return CalendarFieldTest;
}(hornet_react_test_1.HornetReactTest));
//lancement des Tests
test_run_1.runTest(new CalendarFieldTest());
