// app/hornet-js-react-components/test/widget/navigation/bread-crumb-dispatcher-mock-liste-partenaires.js


var menu = [{
    text: "navigation.welcome",
    url: "/accueil",
    visibleDansMenu: true,
    visibleDansPlan: true
}, {
    text: "navigation.contact",
    url: "/messages",
    visibleDansMenu: true,
    visibleDansPlan: true
}, {
    text: "navigation.partenaire.gestionDesPartenaires",
    submenu: [{
        text: "navigation.partenaire.listeViaRecherche",
        url: "/partenaires",
        visibleDansMenu: true,
        visibleDansPlan: true
    }]
}, {
    text: "navigation.accessibilite",
    url: "/accessibilite",
    visibleDansPlan: true
}];

var messages = {
    navigation: {
        planApplication: "Plan de l'application",
        welcome: "Accueil",
        accessibilite: "Accessibilité",
        contact: "Contact",
        planAppli: "Plan de l'application",
        aide: "Aide",
        partenaire: {
            gestionDesPartenaires: "Gestion des partenaires",
            listeViaRecherche: "Liste (via recherche)"
        },
        administration: {
            label: "Administration",
            secteurs: "Secteurs",
            listeSecteurs: "Liste des secteurs"
        }
    }
};

var i18n = require("hornet-js-core/src/i18n/i18n-fluxible-plugin").i18n;

var DispatcherListe = {
    getStore: function()  {
        return {
            getConfigMenu: function()  {
                return menu
            },
            getCurrentUrl: function()  {
                return "/partenaires"
            },
            getCurrentUrlWithoutContextPath: function()  {
                return "/partenaires"
            },
            addChangeListener: function()  {
                return {
                    results: menu,
                    currentUrl: "/partenaires"
                }
            }

        }
    },
    executeAction: function()  {
    },
    locale: "fr-FR",
    i18n: function(keysString)  {
        return i18n(messages)(keysString);
    }
};

module.exports = DispatcherListe;
