// app/hornet-js-react-components/test/widget/navigation/bread-crumb-dispatcher-mock.js


var menu = [{
    text: "navigation.partenaire.listeViaRecherche",
    url: "/partenaires",
    visibleDansMenu: true,
    visibleDansPlan: true
}];

var i18n = require("hornet-js-core/src/i18n/i18n-fluxible-plugin").i18n;

var messages = {
    navigation: {
        partenaire: {
            listeViaRecherche: "Liste (via recherche)"
        }
    }
};

var Dispatcher = {
    getStore: function()  {
        return {
            getConfigMenu: function()  {
                return menu
            },
            getCurrentUrl: function()  {
                return "/partenaires"
            },
            getCurrentUrlWithoutContextPath: function()  {
                return "/partenaires"
            },
            addChangeListener: function()  {
                return {
                    results: menu,
                    currentUrl: "/partenaires"
                }
            }

        }
    },
    executeAction: function()  {
    },
    locale: "fr-FR",
    i18n: function(keysString)  {
        return i18n(messages)(keysString);
    }
};

module.exports = Dispatcher;
