// app/hornet-js-react-components/src/widget/navigation/menu.js
"use strict";
/**
 * Copyright ou © ou Copr. Ministère de l'Europe et des Affaires étrangères (2017)
 * <p/>
 * pole-architecture.dga-dsi-psi@diplomatie.gouv.fr
 * <p/>
 * Ce logiciel est un programme informatique servant à faciliter la création
 * d'applications Web conformément aux référentiels généraux français : RGI, RGS et RGAA
 * <p/>
 * Ce logiciel est régi par la licence CeCILL soumise au droit français et
 * respectant les principes de diffusion des logiciels libres. Vous pouvez
 * utiliser, modifier et/ou redistribuer ce programme sous les conditions
 * de la licence CeCILL telle que diffusée par le CEA, le CNRS et l'INRIA
 * sur le site "http://www.cecill.info".
 * <p/>
 * En contrepartie de l'accessibilité au code source et des droits de copie,
 * de modification et de redistribution accordés par cette licence, il n'est
 * offert aux utilisateurs qu'une garantie limitée.  Pour les mêmes raisons,
 * seule une responsabilité restreinte pèse sur l'auteur du programme,  le
 * titulaire des droits patrimoniaux et les concédants successifs.
 * <p/>
 * A cet égard  l'attention de l'utilisateur est attirée sur les risques
 * associés au chargement,  à l'utilisation,  à la modification et/ou au
 * développement et à la reproduction du logiciel par l'utilisateur étant
 * donné sa spécificité de logiciel libre, qui peut le rendre complexe à
 * manipuler et qui le réserve donc à des développeurs et des professionnels
 * avertis possédant  des  connaissances  informatiques approfondies.  Les
 * utilisateurs sont donc invités à charger  et  tester  l'adéquation  du
 * logiciel à leurs besoins dans des conditions permettant d'assurer la
 * sécurité de leurs systèmes et ou de leurs données et, plus généralement,
 * à l'utiliser et l'exploiter dans les mêmes conditions de sécurité.
 * <p/>
 * Le fait que vous puissiez accéder à cet en-tête signifie que vous avez
 * pris connaissance de la licence CeCILL, et que vous en avez accepté les
 * termes.
 * <p/>
 * <p/>
 * Copyright or © or Copr. Ministry for Europe and Foreign Affairs (2017)
 * <p/>
 * pole-architecture.dga-dsi-psi@diplomatie.gouv.fr
 * <p/>
 * This software is a computer program whose purpose is to facilitate creation of
 * web application in accordance with french general repositories : RGI, RGS and RGAA.
 * <p/>
 * This software is governed by the CeCILL license under French law and
 * abiding by the rules of distribution of free software.  You can  use,
 * modify and/ or redistribute the software under the terms of the CeCILL
 * license as circulated by CEA, CNRS and INRIA at the following URL
 * "http://www.cecill.info".
 * <p/>
 * As a counterpart to the access to the source code and  rights to copy,
 * modify and redistribute granted by the license, users are provided only
 * with a limited warranty  and the software's author,  the holder of the
 * economic rights,  and the successive licensors  have only  limited
 * liability.
 * <p/>
 * In this respect, the user's attention is drawn to the risks associated
 * with loading,  using,  modifying and/or developing or reproducing the
 * software by the user in light of its specific status of free software,
 * that may mean  that it is complicated to manipulate,  and  that  also
 * therefore means  that it is reserved for developers  and  experienced
 * professionals having in-depth computer knowledge. Users are therefore
 * encouraged to load and test the software's suitability as regards their
 * requirements in conditions enabling the security of their systems and/or
 * data to be ensured and,  more generally, to use and operate it in the
 * same conditions as regards security.
 * <p/>
 * The fact that you are presently reading this means that you have had
 * knowledge of the CeCILL license and that you accept its terms.
 *
 */
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
/**
 * hornet-js-react-components - Ensemble des composants web React de base de hornet-js
 *
 * @author MEAE - Ministère de l'Europe et des Affaires étrangères
 * @version v5.1.1
 * @link git+https://github.com/diplomatiegouvfr/hornet-js.git
 * @license CECILL-2.1
 */
var React = require("react");
var navigation_utils_1 = require("hornet-js-components/src/utils/navigation-utils");
var hornet_component_1 = require("hornet-js-react-components/src/widget/component/hornet-component");
var menu_navigation_1 = require("hornet-js-react-components/src/widget/navigation/menu-navigation");
var _ = require("lodash");
var classnames = require("classnames");
var classNames = require("classnames");
var key_codes_1 = require("hornet-js-components/src/event/key-codes");
var menu_info_accessibilite_1 = require("hornet-js-react-components/src/widget/navigation/menu-info-accessibilite");
var hornet_js_utils_1 = require("hornet-js-utils");
var layout_switcher_1 = require("hornet-js-react-components/src/widget/screen/layout-switcher");
var menu_link_1 = require("hornet-js-react-components/src/widget/navigation/menu-link");
var showIconSize = 720;
var Menu = /** @class */ (function (_super) {
    tslib_1.__extends(Menu, _super);
    function Menu(props, context) {
        var _this = _super.call(this, props, context) || this;
        _this.state.items = _this.props.configMenu ? navigation_utils_1.NavigationUtils.getFilteredConfigNavigation(_.cloneDeep(_this.props.configMenu), _this.user) : navigation_utils_1.NavigationUtils.getFilteredConfigNavigation(navigation_utils_1.NavigationUtils.getConfigMenu(), _this.user);
        if (_this.props.closeOnClickOutside)
            _this.handleMenuOutsideClick.bind(_this);
        _this.state.isMenuActive = false;
        _this.handleLayoutExpand();
        _this.listen(layout_switcher_1.UPDATE_PAGE_EXPAND_MENU, _this.handleUpatePageMenuExpand);
        _this.listen(menu_link_1.MENU_LINK_ACTIVATED, _this.handleMajLink);
        return _this;
    }
    /**
     * Change la taille du menu lorque la taille de l'écran est modifiée
     */
    Menu.prototype.handleResize = function () {
        var show = window.innerWidth > showIconSize;
        this.setState({ shouldShowIconInfo: show });
    };
    /**
     * @inheritDoc
     */
    Menu.prototype.componentDidMount = function () {
        window.addEventListener("resize", this.handleResize);
    };
    /**
     * @inheritDoc
     */
    Menu.prototype.componentWillUnmount = function () {
        window.removeEventListener('resize', this.handleResize);
        this.remove(menu_link_1.MENU_LINK_ACTIVATED, this.handleMajLink);
        this.remove(layout_switcher_1.UPDATE_PAGE_EXPAND_MENU, this.handleUpatePageMenuExpand);
    };
    /**
     * Gestion du clic clavier
     * @param event
     */
    Menu.prototype.handleKeyDown = function (event) {
        /*echap*/
        var keyCode = event.keyCode;
        switch (keyCode) {
            case key_codes_1.KeyCodes.ESCAPE:
                this.handleToggleMenu();
                this.burgerIcon.focus();
                break;
            /*tab si l'element focus n'est pas dans le menu, on ferme celui-ci*/
            case key_codes_1.KeyCodes.TAB:
                if (this.state.closeOnTabOutside) {
                    var focus_1 = document.activeElement;
                    var menuMain = document.getElementById("menu-main");
                    /*le changement de focus passe par le body*/
                    var body = document.body;
                    if (menuMain && focus_1 && focus_1 != menuMain && focus_1 != body) {
                        var inMenu = menuMain.contains(focus_1);
                        if (!inMenu) {
                            this.handleToggleMenu();
                        }
                    }
                }
                break;
        }
    };
    /**
     * @inheritDoc
     */
    Menu.prototype.render = function () {
        var classname = {};
        if (typeof window != "undefined") {
            classname = {
                "vertical": this.state.vertical && (window.innerWidth > showIconSize) ? true : false,
                "menuHaut": true
            };
        }
        else {
            classname = {
                "vertical": this.state.vertical,
                "menuHaut": true
            };
        }
        var menuAttributes = {};
        menuAttributes.className = classNames(classname);
        return (React.createElement("nav", tslib_1.__assign({ id: "menu-main", role: "navigation", "aria-label": this.i18n("menu.mainMenu"), "data-title": "Menu" }, menuAttributes),
            this.state.items.length > 0 ? this.renderMenuButton() : null,
            this.state.isMenuActive ? this.renderNavigationMenu() : this.renderEmptyNavigationMenu()));
    };
    /**
     * Rendu du bouton burger du menu
     * @returns {any}
     */
    Menu.prototype.renderMenuButton = function () {
        var _this = this;
        var buttonClasses = {
            toggler: true,
            "is-active": this.state.isMenuActive
        };
        return (React.createElement("button", { className: classnames(buttonClasses), onClick: this.handleToggleMenu.bind(this), "aria-controls": "menu-container", "aria-expanded": this.state.isMenuActive, "aria-haspopup": true, ref: function (button) {
                _this.burgerIcon = button;
            }, type: "button" },
            React.createElement("span", null, this.state.isMenuActive ? this.i18n("menu.closeMenu") : this.i18n("menu.openMenu"))));
    };
    /**
     * Rendu du menu vide (permet de gérer des transition css sur le menu)
     * @returns {any}
     */
    Menu.prototype.renderEmptyNavigationMenu = function () {
        return (React.createElement("div", { className: "menu-container", id: "menu-container" },
            React.createElement("div", { className: "menuEmpty" }),
            React.createElement("div", { className: "menu-content " + this.state.classNameExpanded, style: { maxWidth: this.state.currentWorkingZoneWidth } })));
    };
    /**
     * Rendu du menu
     * @returns {any}
     */
    Menu.prototype.renderNavigationMenu = function () {
        if (this.state.isMenuActive) {
            var body = document.body;
            body.classList.add("noscroll");
        }
        var shouldShowIconInfo = this.state.shouldShowIconInfo;
        if (typeof shouldShowIconInfo == "undefined") {
            shouldShowIconInfo = window.innerWidth > showIconSize;
        }
        return (React.createElement("div", { className: "menu-container", id: "menu-container" },
            this.props.closeOnClickOutside ? React.createElement("div", { className: "menu-overlay", onClick: this.handleToggleMenu.bind(this) }) : null,
            React.createElement("div", { className: "menu-content " + this.state.classNameExpanded, style: { maxWidth: this.state.currentWorkingZoneWidth } },
                this.state.showIconInfo && shouldShowIconInfo ? React.createElement(menu_info_accessibilite_1.MenuInfoAccessibilite, null) : null,
                React.createElement(menu_navigation_1.MenuNavigation, { items: this.state.items, isVisible: this.state.isMenuActive, closeMenu: this.handleToggleMenu.bind(this), dataPassThru: this.props.dataPassThru, closeOnLinkClick: this.props.closeOnLinkClick }))));
    };
    /**
     * Action permettant d'afficher/masquer le menu
     */
    Menu.prototype.handleToggleMenu = function () {
        var body = document.body;
        body.classList.remove("noscroll");
        var active = !this.state.isMenuActive;
        if (!active) {
            document.removeEventListener("keydown", this.handleKeyDown, false);
        }
        else {
            document.addEventListener("keydown", this.handleKeyDown, false);
        }
        if (this.state.onToggleClick) {
            this.state.onToggleClick();
        }
        this.setState({ isMenuActive: active });
    };
    Menu.prototype.componentDidUpdate = function () {
        //gestion de levent click ua chargement de la page
        if (typeof document !== undefined && this.props.closeOnClickOutside) {
            if (!this.state.isMenuActive) {
                document.removeEventListener("click", this.handleMenuOutsideClick, false);
            }
            else {
                document.addEventListener("click", this.handleMenuOutsideClick, false);
            }
        }
    };
    /**
     *
     * @param e Event
     */
    Menu.prototype.handleMenuOutsideClick = function (e) {
        e.stopPropagation();
        return false;
    };
    Menu.prototype.handleLayoutExpand = function () {
        // préparation de la taille pour le layout expanding
        var maxWidth;
        var classNameExpanded = "mainLayoutClassNameExpanded";
        if (!(hornet_js_utils_1.Utils.appSharedProps.get("isExpandedLayout"))) {
            maxWidth = this.state.workingZoneWidth;
            classNameExpanded = "mainLayoutClassName";
        }
        this.state.currentWorkingZoneWidth = maxWidth;
        this.state.classNameExpanded = classNameExpanded;
    };
    Menu.prototype.handleMajLink = function () {
        this.setState({ majlink: true });
    };
    Menu.prototype.handleUpatePageMenuExpand = function () {
        this.handleLayoutExpand();
        this.setState({ maj: true });
    };
    Menu.defaultProps = {
        closeOnLinkClick: true,
        closeOnClickOutside: true,
        closeOnTabOutside: true
    };
    return Menu;
}(hornet_component_1.HornetComponent));
exports.Menu = Menu;
